crates/store/re_query2/src/range.rs

```diff
@@ -0,0 +1,317 @@
+        } = self;
+
+        let mut strings: Vec<String> = Vec::new();
+
+        strings.push(format!(
+            "{} ({})",
+            cache_key.timeline.typ().format_range_utc(self.time_range()),
+            re_format::format_bytes(chunks.total_size_bytes() as _),
+        ));
+
+        if strings.is_empty() {
+            return f.write_str("<empty>");
+        }
+
+        f.write_str(&strings.join("\n").replace("\n\n", "\n"))
+    }
+}
+
+pub struct RangeCachedChunk {
+    pub chunk: Chunk,
+
+    /// When a `Chunk` gets cached, it is pre-processed according to the current [`CacheKey`],
+    /// e.g. it is time-sorted on the appropriate timeline.
+    ///
+    /// In the happy case, pre-processing a `Chunk` is a no-op, and the cached `Chunk` is just a
+    /// reference to the real one sitting in the store.
+    /// Otherwise, the cached `Chunk` is a full blown copy of the original one.
+    pub resorted: bool,
+}
+
+impl SizeBytes for RangeCachedChunk {
+    #[inline]
+    fn heap_size_bytes(&self) -> u64 {
+        let Self { chunk, resorted } = self;
+
+        if *resorted {
+            // The chunk had to be post-processed for caching.
+            // Its data was duplicated.
+            Chunk::heap_size_bytes(chunk)
+        } else {
+            // This chunk is just a reference to the one in the store.
+            // Consider it amortized.
+            0
+        }
+    }
+}
+
+impl SizeBytes for RangeCache {
+    #[inline]
+    fn heap_size_bytes(&self) -> u64 {
+        let Self {
+            cache_key,
+            chunks,
+            pending_invalidations,
+        } = self;
+
+        cache_key.heap_size_bytes()
+            + chunks.heap_size_bytes()
+            + pending_invalidations.heap_size_bytes()
+    }
+}
+
+impl RangeCache {
+    /// Queries cached range data for a single component.
+    pub fn range(
+        &mut self,
+        store: &ChunkStore,
+        query: &RangeQuery,
+        entity_path: &EntityPath,
+        component_name: ComponentName,
+    ) -> Vec<Chunk> {
+        re_tracing::profile_scope!("range", format!("{query:?}"));
+
+        debug_assert_eq!(query.timeline(), self.cache_key.timeline);
+
+        // First, we forward the query as-is to the store.
+        //
+        // It's fine to run the query every time -- the index scan itself is not the costly part of a
+        // range query.
+        //
+        // For all relevant chunks that we find, we process them according to the [`CacheKey`], and
+        // cache them.
+
+        let raw_chunks = store.range_relevant_chunks(query, entity_path, component_name);
+        for raw_chunk in &raw_chunks {
+            self.chunks
+                .entry(raw_chunk.id())
+                .or_insert_with(|| RangeCachedChunk {
+                    chunk: raw_chunk
+                        // Pre-sort the cached chunk according to the cache key's timeline.
+                        .sorted_by_timeline_if_unsorted(&self.cache_key.timeline)
```

Comment:
I believe there is an inherent inefficiency here that should eventually be optimized. It's mitigated by the fact that we cache the result, but still seems wasteful.

When the same chunk contains 2 or more of our components that are going to be used in the same top-level joined range query, we are going to end up sorting that chunk multiple times, using the same swaps. Once for each executation of range(...) with a different component.

Worse, as far as I can tell, each time we call `.sorted_by_timeline_if_unsorted` we actually end up sorting ALL of the internal components stored in that chunk, even though we ultimately densify for just a single component.

Reply:
> When the same chunk contains 2 or more of our components that are going to be used in the same top-level joined range query, we are going to end up sorting that chunk multiple times, using the same swaps. Once for each executation of range(...) with a different component.

That's correct, with one caveat though: only on the unhappy path.

On the happy path, sorting is a no-op both in space and time, so this is not an issue.

On the unhappy path, it is inefficient in _both_ space and time:
1. We're re-sorting for each component.
2. We're caching a _different copy_ of the sorted chunks each time.

I'll open and link to an issue for that.

> Worse, as far as I can tell, each time we call `.sorted_by_timeline_if_unsorted` we actually end up sorting ALL of the internal components stored in that chunk, even though we ultimately densify for just a single component.

You're right, there's no good reason not to call `densify` first!